### PR TITLE
Option to create manifest and requirements.txt

### PR DESCRIPTION
**Pull request by mmarchetti**
### Description

This PR adds a button to the publishing dialog to create a manifest and requirements.txt. This enables publishing notebooks via the new git support in Connect.

Additionally, an existing requirements.txt will be honored if present (it won't be overwritten, and will be used in preference to inspecting the environment with `pip freeze`). Currently, the requirements file must include `nbconvert` and `ipykernel` or packages which depend on them in order for rendering to work.

Connected to #15350

### Testing Notes / Validation Steps
* Use the Create Manifest button to create a manifest and requirements file. Try clicking the links to edit them (especially the requirements file). You can also see them in the Jupyter file manager. Deploy, and see that the requirements you specified are included in the requirements build.
* Remove the manifest and create again. It should create the manifest but leave your requirements.txt file alone.
* Push the files (notebook, requirements, manifest) to a git repo and deploy to Connect via git. The deployment should succeed and again honor your requirements.txt file.

### CODE REVIEW

**Comment by slopp** [comment]
@mmarchetti can you walk me through the workflow here? Is the idea that users would pull up the dialog., check the box to create a manifest, publish the content, and then subsequently have the artifacts needed to commit to Git? Then RSC could be setup to auto-deploy a second version of that content? 

I was thinking as a user I'd want a way to create the manifest and requirements file separate from the push button publish process, because: 

1. I don't want two versions of the content on RSC, one push-button deployed and one Git deployed 
2. I don't want to fire up the push button extension to prepare for my Git commit 

Perhaps I am missing something since I didn't fire this up?

**Comment by mmarchetti** [comment]
@slopp the workflow would be to click the icon in Jupyter which opens the Publish dialog, then click the Create Manifest button. Optionally, use the links to open the created `manifest.json` and `requirements.txt` files within Jupyter. At that point you can close the dialog, you don't need to publish. Commit `manifest.json` and `requirements.txt` to git along with the notebook, push to a git server, and you can deploy via git in Connect.

**Comment by mmarchetti** [comment]
I am interested in getting feedback on this. It's certainly possible to do other things, for example, we could turn the main Jupyter blue button into some kind of dropdown menu with choices like Publish and Create Manifest (I don't really like that idea because it adds a click to the primary publishing flow). Or, we could add a second button to the notebook UI, parallel to the Publish button, which would show a Create Manifest dialog.

**Comment by mmarchetti** [comment]
We're carrying a separate issue to split some functionality out of rsconnect-jupyter into a separate library package with fewer dependencies (not Jupyter). This could also provide a CLI tool to do that and potentially other tasks.

**Comment by mmarchetti** [comment]
Took a shot at making a separate dialog for this, with a separate button to trigger without having to open the publishing dialog.

Buttons in Jupyter toolbar:
Screen Shot 2019-10-02 at 11 11 29 AM

The dialog (after a successful creation):

Screen Shot 2019-10-02 at 11 10 05 AM

**Comment by slopp** [comment]
I think this is great Mike! It makes a lot more sense to me. I would like to get @koverholt 's second opinion. Only nit: I think it would be great to add a link somewhere on this second button to the user guide info on Git deployments.

**Comment by koverholt** [comment]
Thanks for the ping. This is good stuff @mmarchetti 👍

I installed the dev version from this PR and tried it out.

The functionality to generate the manifest and `requirements.txt` is great. That worked nicely for me. The links that open the text files are great. I did find it slightly confusing that the "Generate Manifest" button generated a manifest *and* a `requirements.txt` file.

I am not a fan of having two separate buttons to work with RSC from Jupyter. Most users (and myself) will not remember what the two different buttons are for and in what order to click them depending on what they are trying to do. And we should not make them distinct actions in the Jupyter UX just because the RStudio IDE has a button for publishing whereas users have to issue a separate R command in the console to generate the manifest.

Instead, I like the approach of the drop-down in RSC for publishing, you click one thing to publish, and it guides you to the correct modal or method depending on what you want to do. I prefer to point a user to one action/button, then help them get to the right path from there, such as:

1) Push-button publishing, then show the appropriate info for them to fill out
2) Git-backed publishing, follow these steps to generate manifest and `requirements.txt`, then commit, then go to RSC and publish from git

**Comment by mmarchetti** [comment]
Switched to a single button with dropdown menu.

Screen Shot 2019-10-03 at 2 24 33 PM

**Comment by robbyshaver** [comment]
I'm glad you settled on the drop down menu for the two paths - makes a ton of sense and is very clear.

File: rsconnect_jupyter/static/connect.js [hunk 1]
```diff
@@ -571,11 +592,13 @@ define([
       open: function() {
         disableKeyboardManagerIfNeeded();
         // TODO add ability to dismiss via escape key
+        var $deploy_err = $('#rsc-deploy-error');
```
Comment (scottmmjackson) [inline rsconnect_jupyter/static/connect.js]:
```suggestion
```

This is unused in this scope

File: rsconnect_jupyter/static/connect.js [hunk 2]
```diff
@@ -685,6 +708,7 @@ define([
           publishModal.find('.help-block').text('');
           var $deploy_err = $('#rsc-deploy-error');
```
Comment (scottmmjackson) [inline rsconnect_jupyter/static/connect.js]: Removing the above `$deploy_err` should suppress the `no-shadow` here.

File: rsconnect_jupyter/static/connect.js [hunk 3]
```diff
@@ -1004,6 +1028,16 @@ define([
     });
   }
 
+  function onMenuClicked($button) {
```
Comment (scottmmjackson) [inline rsconnect_jupyter/static/connect.js]:
```suggestion
  function onMenuClicked() {
```

not used

File: rsconnect_jupyter/static/connect.js [hunk 4]
```diff
@@ -42,10 +42,31 @@ define([
     Jupyter.toolbar.add_buttons_group([actionName]);
```
Comment (scottmmjackson) [inline rsconnect_jupyter/static/connect.js]: ESLint autofix will change the double-quote strings to singlequote

File: rsconnect_jupyter/static/connect.js [hunk 5]
```diff
@@ -1037,6 +1073,78 @@ define([
...
+      // modal is active
+      keyboard_manager: Jupyter.notebook.keyboard_manager,
+
+      title: 'Create Manifest',
+      body: [
+        '<p>',
+        '    Click Create Manifest to create the manifest.json and requirements.txt',
+        '    files needed for publishing to RStudio Connect via git.',
+        '    Note: requirements.txt will be created only if it does not already exist.',
+        '</p>',
```
Comment (scottmmjackson) [inline rsconnect_jupyter/static/connect.js]: The note should probably be a second paragraph.

"Create Manifest" should probably be quoted to isolate it from the rest of the verbiage.

This seems like a pretty good summary of what's going on but I couldn't help but feel like I didn't understand why there was a dialog here. Perhaps we could explain to the user what the consequences of doing this if they don't actually need it are?

Comment (scottmmjackson) [reply rsconnect_jupyter/static/connect.js]: I'd be interested to see a sample of manifest/requirements generation failing. I can reproduce a success locally, but I'm not sure I understand what failure looks like.

Success 1:

Screen Shot 2019-10-06 at 3 39 51 PM

Shamelessly spamming the "Create Manifest" button:

Screen Shot 2019-10-06 at 3 40 18 PM

Comment (mmarchetti) [reply rsconnect_jupyter/static/connect.js]: This is the expected behavior if you click the button a second time. Seems like we need some feedback that requirements.txt already existed and was not overwritten.

**Comment by mmarchetti** [comment]
That's a good question. We're extra careful with requirements.txt because it's very common to have a pre-existing one, but doing both the same might be easier to understand. There really isn't anything that changes in the manifest for a notebook over time, maybe the version of pip that's installed.

**Comment by kgartland-rstudio** [comment]
When creating a manifest, the `files` field is empty:

`  "files": {},`

Should this instead include the `requirements.txt` and notebook file?

**Comment by mmarchetti** [comment]
We're currently not filling those in, because the manifest can be reused for subsequent git deployments. Otherwise they would have to remember to regenerate and commit the manifest after they edit the notebook, before they could deploy again.

**Comment by kgartland-rstudio** [comment]
Verified.  
Created #15494 to change the behavior of manifest.json generation when that file already exists in current directory.

**Pull request merged**